src/Voron/GlobalPrefetchingBehavior.cs

```diff
@@ -0,0 +1,105 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using Sparrow;
+using Sparrow.Collections;
+using Sparrow.Logging;
+using Sparrow.Utils;
+using Voron.Platform.Win32;
+
+namespace Voron.Impl
+{
+    public class GlobalPrefetchingBehavior
+    {
+        public unsafe struct PrefetchCommand
+        {
+            public void* Base;
+            public long Length;
+        }
+
+        private const string PrefetchingThreadName = "Voron Global Prefetching Thread";
+
+        internal static readonly Lazy<GlobalPrefetchingBehavior> GlobalPrefetcher = new Lazy<GlobalPrefetchingBehavior>(() =>
+        {
+            var prefetcher = new GlobalPrefetchingBehavior();
+            var thread = new Thread(prefetcher.VoronPrefetcher)
+            {
+                IsBackground = true,
+                Name = PrefetchingThreadName
+            };
+            thread.Start();
+            return prefetcher;
+        });
+
+        private readonly Logger _log = LoggingSource.Instance.GetLogger<GlobalPrefetchingBehavior>("Global Prefetcher");
+         
+        public readonly SingleConsumerRingBuffer<PrefetchCommand> CommandQueue = new SingleConsumerRingBuffer<PrefetchCommand>(128);
```

Comment:
Yes we do, because the one that is going to pay for it is the caller... The whole point of the optimization is to move the slow parts outside of the critical path, even at the expense of losing a few hits here and there.

Reply:
Using `BlockingCollection`